Umut Demirhan 

Student ID: 46739106

# Business Analytics - Assignment 2  

**Assignment Points**: 100  
**Submission**: Provide your answers in this notebook and submit it via iLearn

- Where a question requires a written (text) solution provide your answer in Markdown in appropriate cells under each question.
- Comment out your print statements unless you are explicitly asked to use the print() function. 
- 5 marks will be deducted for printed outputs that are not asked for.

### About the Assignment

- Assignment 2 extends Assignment 1 on credit card applications. 


- For this assignment there are two files in the `data` folder `credit_record.csv` and `application_record.csv` where bank clients are related by the `ID` column.
- In `credit_record.csv` we have the following variables

| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number      |         |
| AMT_INCOME_TOTAL   | Annual income  |  |
| NAME_INCOME_TYPE   | Income Source |  |
| NAME_EDUCATION_TYPE   | Level of Education  |  |
| CODE_GENDER   | Applicant's Gender   |  |
| FLAG_OWN_CAR | Car Ownership |  | 
| CNT_CHILDREN | Number of Children | |
| FLAG_OWN_REALTY | Real Estate Ownership | | 
| NAME_FAMILY_STATUS | Relationship Status | | 
| NAME_HOUSING_TYPE | Housing Type | | 
| DAYS_BIRTH | No. of Days | Count backwards from current day (0), -1 means yesterday
| DAYS_EMPLOYED | No. of Days | Count backwards from current day(0). If positive, it means the person is currently unemployed.
| FLAG_MOBIL | Mobile Phone Ownership | | 
| FLAG_WORK_PHONE | Work Phone Ownership | | 
| FLAG_PHONE | Landline Phone Ownership | | 
| FLAG_EMAIL | Landline Phone Ownership | | 
| OCCUPATION_TYPE | Occupation | | 
| CNT_FAM_MEMBERS | Count of Family Members | |



- In `credit_record.csv` we have the following variables


| Feature Name         | Explanation     | Additional Remarks |
|--------------|-----------|-----------|
| ID | Randomly allocated client number | |
| MONTHS_BALANCE | Number of months in the past from now when STATUS is measured | 0 = current month, -1 = last month, -2 = two months ago, etc.|
| STATUS | Number of days a payment is past due | 0: 1-29 days past due 1: 30-59 days past due 2: 60-89 days overdue 3: 90-119 days overdue 4: 120-149 days overdue 5: Overdue or bad debts, write-offs for more than 150 days C: paid off that month X: No loan for the month |

---
---

### Task 1: Reading, Summarising and Cleaning Data (Total Marks: 30)



**Question 1.** 

1. Import the `application_record.csv` and `credit_record.csv` files from `data` folder into pandas DataFrames named `df_application` and `df_credit`, respectively. (1 mark)

2. How many rows are there in `df_application` and `df_credit`, respectively? Answer using both print() function and in Markdown text. (1 mark)

3. How many unique bank clients are there in `df_application` and `df_credit`? Answer using both print() function and in Markdown text. (1 mark)

4. Add the records from `df_credit` to `df_application` by merging the data from the two DataFrames on the `ID` column, and output the joint data into a new DataFrame named `df`. Hint: Use `merge` function from pandas by setting `how` parameter to `inner` (4 marks) 

5. How many rows and how many unique clients are there in `df`? (1 mark)

6. How are multiple rows for each `ID` in `df` different? Answer in Markdown text. (2 mark) 

(10 marks)


In [1]:
import pandas as pd
import numpy as np

# Answer 1.1
df_application = pd.read_csv('data/application_record.csv')
df_credit = pd.read_csv('data/credit_record.csv')

# Amswer 1.2
print("Number of rows in df_application: ",   len(df_application))
print("Number of rows in df_credit: ", len(df_credit))

# Answer 1.3
print("Number of unique bank clients in df_application: ", df_application['ID'].nunique())
print("Number of unique bank clients in df_credit: ", df_credit['ID'].nunique())

# Answer 1.4
df = pd.merge(df_application, df_credit, on='ID', how='inner')

# Answer 1.5 
#Number of rows in df:
#df.shape[0]
#Number of unique clients in df:
#df['ID'].nunique()
#df['MONTHS_BALANCE'].unique()


Number of rows in df_application:  438557
Number of rows in df_credit:  1048575
Number of unique bank clients in df_application:  438510
Number of unique bank clients in df_credit:  45985


###### Answer 1.2 
* Number of rows in 'df_application' is 438557 and, number of rows in 'df_credit' is 1048575
###### Answer 1.3
* Number of unique bank clients in  'df_application' is 438510 and number of unique bank clients in 'df_credit' is 45985
###### Answer 1.5
* Number of rows in 'df' is 777715 and, number of unique clients in 'df' is 36457

###### Answer 1.6
* Because we merged the two data frames, the multiple rows for each ID in df indicate that there are multiple records for each client. Therefore, if we delete the 'month balance' and 'status' columns, then 'df' will have many duplicate rows.

---

**Question 2.**

1. Change the values of `STATUS` in `df` according to the following mapping: {C, X, 0} -> 0 and {1, 2, 3, 4, 5} -> 1 making sure that the new values of 0 and 1 are encoded as integers. (2 marks)
2. Create a new numpy array named `list_of_defaults` containing *unique* `ID` numbers for the clients who have `STATUS` = 1 in any of the last 12 months in the dataset. (2 marks) 
3. Create a new DataFrame called `df_final` that contains the rows of `df` for which the `ID` are in `list_of_defaults`, keeping only one row for each `ID` (i.e. eliminate rows with duplicate `ID`s while keeping the first duplicate row). How many rows do you have in `df_final`? Answer using both print() function and in Markdown text. (Hint: find out about `isin()` function in pandas.) (2 marks)
4. Add a new column `y = 1` for all the rows in `df_final`. (1 marks)
5. Increase `df_final` to a total of 4,000 rows by adding rows from `df` with unique `ID`s (nonduplicated `ID`s) which are not in `list_of_defaults`. To do this start adding the rows from the beginning of `df`. (Hint: learn what `~`, i.e. tilde sign, does in pandas). (2 marks) 
6. Fill the missing values of `y` in `df_final` with zeros. Remove `STATUS` and `MONTHS_BALANCE` from `df_final`. How many clients with  overdue payments of more than 29 days and how many clients with less than 29 days overdue payments are there in `df_final`? Answer using both print() function and in Markdown text.(1 mark)

(10 marks)

In [2]:
# Answer 2.1
df['STATUS'] = df['STATUS'].replace(['C', 'X', 0], 0).replace([1, 2, 3, 4, 5], 1).astype(int)
# Answer 2.2
list_of_defaults = df[df['MONTHS_BALANCE'] >= -12]['ID'][df['STATUS'] == 1].unique()
# Answer 2.3
df_final = df[df['ID'].isin(list_of_defaults)].drop_duplicates(subset='ID', keep='first')
print("Number of rows in df_final:", len(df_final))
# Answer 2.4
df_final['y'] = 1
# Answer 2.5
df_nondefault = df[~df['ID'].isin(list_of_defaults)].drop_duplicates(subset='ID', keep='first')    
df_final = pd.concat([df_final, df_nondefault.iloc[:4000-df_final.shape[0]]])
# Answer 2.6
df_final['y'] = df_final['y'].fillna(0)
df_final = df_final.drop(['STATUS', 'MONTHS_BALANCE'], axis=1)
overdue_29 = df_final[df_final['y'] > 0]['ID'].nunique()
under_29 = df_final[df_final['y'] == 0]['ID'].nunique()
print("Number of clients with overdue payments of more than 29 days:", overdue_29)
print("Number of clients with overdue payments of less than or equal to 29 days:", under_29)

Number of rows in df_final: 1721
Number of clients with overdue payments of more than 29 days: 1721
Number of clients with overdue payments of less than or equal to 29 days: 2279


###### Answer 2.3
* Number of rows in 'df_final' is 1721
###### Answer 2.6
* Number of clients with overdue payments of more than 29 days is 1721
* Number of clients with overdue payments of less than or equal to 29 days is 2279

<hr style="width:25%;margin-left:0;"> 

**Question 3**. 
- Delete `ID` column from `df_final` (1 marks)
- Of the remaining variables in `df_final` and assuming that `NAME_EDUCATION_TYPE` is the only ordinal variable, how many variable are of numeric and nominal types? Provide lists of all numeric and nominal variables. (6)
- Using an appropriate function find and comment on the missing values in `df_final`, i.e. how many variables and how many observations? (3 marks)   
(10 marks)

In [3]:
# Answer 3.1
df_final.drop('ID', axis=1, inplace=True)
# Answer 3.2
num_var = list(df_final.select_dtypes(include=['int64', 'float64']).columns)
nom_var = list(df_final.select_dtypes(include=['object', 'category']).columns)
nom_var.remove('NAME_EDUCATION_TYPE')
#len(num_var)
#len (nom_var)
# Answer 3.3
# df_final.isna().sum()

###### Answer 3.2

* There are 10 numeric varible including newly created variable 'y'
    * List of  all numeric variables;
['CNT_CHILDREN',|
 'AMT_INCOME_TOTAL',
 'DAYS_BIRTH',
 'DAYS_EMPLOYED',
 'FLAG_MOBIL',
 'FLAG_WORK_PHONE',
 'FLAG_PHONE',
 'FLAG_EMAIL',
 'CNT_FAM_MEMBERS',
 'y']

* There are 7 nominal varible exluding 'NAME_EDUCATION_TYPE' as it is ordinal variable
    * List of  all nominal variables;
['CODE_GENDER',
 'FLAG_OWN_CAR',
 'FLAG_OWN_REALTY',
 'NAME_INCOME_TYPE',
 'NAME_FAMILY_STATUS',
 'NAME_HOUSING_TYPE',
 'OCCUPATION_TYPE']

###### Answer 3.3

* Three variables have missing values which are 'CNT_CHILDREN', 'NAME_EDUCATION_TYPE', and 'OCCUPATION_TYPE' 
* They have 74, 1831 and 1152 missing observations respectively.

---
---

### Task 2: Imputing missing values and dealing with categorical features (Total Marks: 30)



**Question 4.** 
- Use an appropriate `pandas` function to impute missing values in `df_final` (10 marks)
    - Be careful when deciding which method to use to replace missing observations 
    - Take into consideration the type of each variable and the best practices we discussed in class/lecture notes
- Briefly explain what you have done and why. (5 marks)

(Total: 15 marks)

In [4]:

# Replace missing values in CNT_CHILDREN column with 0
df_final['CNT_CHILDREN'].fillna(0, inplace=True)

# Replace missing values in NAME_EDUCATION_TYPE and OCCUPATION_TYPE columns with the mode
df_final['NAME_EDUCATION_TYPE'].fillna(df_final['NAME_EDUCATION_TYPE'].mode()[0], inplace=True)
df_final['OCCUPATION_TYPE'].fillna(df_final['OCCUPATION_TYPE'].mode()[0], inplace=True)

###### Answer 4
* For the 'CNT_CHILDREN' column, I replaced the missing values with 0, assuming that clients with missing values have no children. 
* For the 'NAME_EDUCATION_TYPE' and 'OCCUPATION_TYPE' columns, I used the mode() method to compute the mode of the non-missing values, and replace the missing values with the mode. Because,  This is a common approach for imputing missing values in categorical variables.


<hr style="width:25%;margin-left:0;"> 

**Question 5**. Convert the values in `NAME_EDUCATION_TYPE` as follows
- Lower secondary -> 1
- Secondary / secondary special -> 2
- Incomplete higher -> 3
- Higher education -> 4


(Total: 5 marks)  

In [5]:
# Answer 5
education_dict = {
    'Lower secondary': 1,
    'Secondary / secondary special': 2,
    'Incomplete higher': 3,
    'Higher education': 4
}

df_final['NAME_EDUCATION_TYPE'] = df_final['NAME_EDUCATION_TYPE'].map(education_dict)

<hr style="width:25%;margin-left:0;"> 

**Question 6**. 

Add dummy variables to `df_final` for all of the nominal features which are currently stored as string (text). 
- Make sure to delete the original variables from the dataframe
- Drop the first column from each set of created dummy variable, i.e. for each feature



(Total: 10 marks)  

In [6]:
for feature in nom_var:
    dummies = pd.get_dummies(df_final[feature], prefix=feature, drop_first=True)
    df_final = pd.concat([df_final, dummies], axis=1)
    df_final.drop(columns=[feature], inplace=True)

---
---

### Task 3 Preparing X and y arrays (Total Marks: 10)

**Question 7**. 

- Create a numpy array named `y` from the `y` column of `df_final` making sure that the values of the array `y` are stored as integers (3 marks)   
- Create a numpy array named `X`  from all the remaining features in `df_final` (2 marks)   

(Total: 5 Marks)

In [7]:
# Answer 7
y = np.array(df_final['y'], dtype=int)

X = np.array(df_final.drop('y', axis=1))

<hr style="width:25%;margin-left:0;"> 

**Question 8**. 

- Use an appropriate scikit-learn library we used in class to create `y_train`, `y_test`, `X_train` and `X_test` by splitting the data into 70% train and 30% test datasets (2.5 marks) 
    - Set random_state to 7 and stratify the subsamples so that train and test datasets have roughly equal proportions of the target's class labels 
- Standardise the data using `StandardScaler` library (2.5 marks)   

(Total: 5 marks) 

In [8]:
# Answer 8
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Splitting the data into 70% train and 30% test datasets, stratified by the target variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=7, stratify=y)

# Standardizing the data using StandardScaler
scaler = StandardScaler()
X_train_std = scaler.fit_transform(X_train)
X_test_std = scaler.transform(X_test)

---
---

### Task 4. Support Vector Classifier and Accuracies (Total Marks: 30)


**Question 9**. 

- Train a Support Vector Classifier on standardised data (3 marks)
    - Use `linear` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- Train another Support Vector Classifier on standardised data (3 marks)
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
    - Compute and print training and test dataset accuracies
- What can you say about the presence of nonlinearities in the dataset? (4 marks)

(Total: 10 marks)  

In [9]:
# Answer 9
from sklearn.svm import SVC

# Training a Support Vector Classifier with linear kernel
svm_linear = SVC(kernel='linear', random_state=7)
svm_linear.fit(X_train_std, y_train)

# Computing training and test dataset accuracies
train_acc_linear = svm_linear.score(X_train_std, y_train)
test_acc_linear = svm_linear.score(X_test_std, y_test)
print("Training accuracy with linear kernel: {:.2f}%".format(train_acc_linear*100))
print("Test accuracy with linear kernel: {:.2f}%".format(test_acc_linear*100))

# Training a Support Vector Classifier with rbf kernel
svm_rbf = SVC(kernel='rbf', random_state=7)
svm_rbf.fit(X_train_std, y_train)

# Computing training and test dataset accuracies
train_acc_rbf = svm_rbf.score(X_train_std, y_train)
test_acc_rbf = svm_rbf.score(X_test_std, y_test)
print("Training accuracy with rbf kernel: {:.2f}%".format(train_acc_rbf*100))
print("Test accuracy with rbf kernel: {:.2f}%".format(test_acc_rbf*100))

Training accuracy with linear kernel: 65.32%
Test accuracy with linear kernel: 64.58%
Training accuracy with rbf kernel: 78.43%
Test accuracy with rbf kernel: 71.42%


#### Answer 9
Based on the provided accuracy scores above, we can conclude that the dataset likely contains non-linearities that cannot be captured by a linear model. The reason is that the accuracy scores obtained by the linear kernel are significantly lower than the scores obtained by the rbf kernel, which is a non-linear kernel. The higher accuracy achieved by the rbf kernel on both the training and test data suggests that it is better at capturing the underlying relationships in the data. Therefore, we can infer that the dataset likely contains non-linear relationships between the features and the target variable.

<hr style="width:25%;margin-left:0;"> 

**Question 10**

- Extract 2 linear principal components from the standardised features using an appropriate `sklearn` library (5 marks)
- Train a Support Vector Classifier on the computed principal components (5 marks) 
    - Use `rbf` kernel and set `random_state` to 7 (don't change any other parameters)
- Compute and print training and test dataset accuracies (5 marks)
- What can you say about the ability of the 2 principal components to compress the information contained in the features matrix `X`, and why? (5 marks)     


(Total: 20 marks)  

In [10]:
# Answer 10
from sklearn.decomposition import PCA

pca = PCA(n_components=2)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

# Train a Support Vector Classifier on the computed principal components using rbf kernel
svc_pca = SVC(kernel='rbf', random_state=7)
svc_pca.fit(X_train_pca, y_train)

train_acc_pca = svc_pca.score(X_train_pca, y_train)
test_acc_pca = svc_pca.score(X_test_pca, y_test)

print('PCA with 2 components:')
print(f'Training accuracy: {train_acc_pca:.4f}')
print(f'Test accuracy: {test_acc_pca:.4f}')


PCA with 2 components:
Training accuracy: 0.5721
Test accuracy: 0.5717


#### Answer 10
Based on the results obtained, it appears that the 2 principal components resulting from the PCA transformation were not able to fully capture the information contained in the features matrix X. The low training and test accuracies obtained with the SVC model trained on these principal components suggest that some important information may have been lost during the dimensionality reduction process.
Therefore, it's possible that more principal components would be needed to fully capture the information contained in the data.

---
---